src/permutations.rs

```diff
@@ -36,19 +36,20 @@ impl<F: Field> MonomialPermutation<F> {
 
 impl<F: Field> Permutation<F> for MonomialPermutation<F> {
     fn permute(&self, builder: &mut GadgetBuilder<F>, x: &Expression<F>) -> Expression<F> {
-        builder.exp(x, &self.n)
+        builder.exponentiation(x, &self.n)
     }
 
     fn inverse(&self, builder: &mut GadgetBuilder<F>, x: &Expression<F>) -> Expression<F> {
         let root_wire = builder.wire();
         let root = Expression::from(root_wire);
-        let exponentiation = builder.exp(&root, &self.n);
+        let exponentiation = builder.exponentiation(&root, &self.n);
         builder.assert_equal(&exponentiation, x);
 
-        // By Fermat's little theorem, x^p = x, so if n divides e, then x^(p / n)^n = x.
-        // Further, since x^(p - 1) = 1, x^((p + (p - 1)*k) / n)^n = x for any positive k,
+        // By Fermat's little theorem, x^p = x mod p, so if n divides p, then x^(p / n)^n = x mod p.
+        // Further, since x^(p - 1) = 1 mod p, x^((p + (p - 1)*k) / n)^n = x mod p for any positive k,
         // provided that n divides p + (p - 1)*k. Thus we start with p, and repeatedly add
         // p - 1 until we find an exponent divisible by n.
+        //TODO: find a solution that isn't O(p)
```

Comment:
If this is unoptimized then `inverse()` will take significantly more time to process than `permute()`.

Reply:
Agreed! Need to look into better algorithms...